pgjdbc/src/main/java/org/postgresql/jdbc/ArrayDecoding.java

```diff
@@ -422,11 +490,66 @@ Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) throw
      * {@inheritDoc}
      */
     @Override
-    Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+    public Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
       return arrayAssistant.buildElement(stringVal);
     }
   }
 
+  private static final class StructTypeArrayDecoder extends AbstractObjectArrayDecoder<Struct[]> {
+
+    private final String typeName;
+
+    StructTypeArrayDecoder(String typeName) {
+      super(PgStruct.class);
+      this.typeName = typeName;
+    }
+
+    @Override
+    public Object parseValue(int length, ByteBuffer bytes, BaseConnection connection) throws SQLException {
+      final PgStruct struct = castNonNull(connection.getTypeInfo().getPgStruct(typeName));
+
+      final byte[] structValue = new byte[length];
+      bytes.get(structValue, 0, length);
+
+      final Tuple row = PgResultSet.fromStructBin(structValue);
+
+      return buildStruct(struct, row, connection, true);
+    }
+
+    @Override
+    public Object parseValue(String stringVal, BaseConnection connection) throws SQLException {
+      final PgStruct struct = castNonNull(connection.getTypeInfo().getPgStruct(typeName));
+      final Charset charset = Charset.forName(connection.getEncoding().name());
+
+      final Tuple row = PgResultSet.fromStruct(stringVal.getBytes(charset));
+
+      return buildStruct(struct, row, connection, false);
+    }
+
+    private Struct buildStruct(PgStruct struct, Tuple row, BaseConnection connection, boolean isBinary) throws SQLException {
+      final Object[] attributes = new Object[struct.getFields().length];
+      for (int i = 0; i < attributes.length; i++) {
+        final PgStructField field = struct.getFields()[i];
+        @SuppressWarnings("rawtypes")
+        final ArrayDecoder decoder = getDecoder(field.getOID(), connection);
+        final byte[] val = row.get(i);
+
+        if (val == null) {
+          attributes[i] = null;
+          continue;
+        }
+        if (isBinary) {
+          final ByteBuffer buffer = ByteBuffer.wrap(val);
+          attributes[i] = decoder.parseValue(val.length, buffer, connection);
+        } else {
+          final Charset charset = Charset.forName(connection.getEncoding().name());
+          attributes[i] = decoder.parseValue(new String(val, charset), connection);
+        }
```

Comment:
This appears to be using the ArrayDecoder to decode singleton fields - which is decidedly NOT the purpose of the ArrayDecoder interface.

However, the concept is reasonable, and seemingly reusable from PgResultSet as well.
Perhaps a FieldEncoder/FieldDecoder pattern should be developed.

Reply:
I totally agree with this as it was also something that I considered when first implementing it, it just seemed that it was the fastest way to try and required the least amount of changes overall in this particular class.

I have added a FieldDecoder interface, can you take a look to see if it is similar with what you had in mind?
I have left those in the ArrayDecoding class as they are only used there but we can consider moving them under a different class.

Note: I didn't develop FieldEncoder-s as Struct binary encoding is not supported due to the small amount of types that actually support binary encoding, it will use the default Object encoder that creates a string array literal. But if you want I can also take a look at how that would look like.